### DISCUSSION TITLE
expose `injectRoute`

### DISCUSSION THREAD

**Discussion by thepassle** [withastro/roadmap#201] [Proposal]
Recently, @FredKSchott made a thread on the discord #feedback-and-suggestions channel that proposes a `injectRoute` API to the integrations `'astro:config:setup'` hook, similar to `injectScript`. This API would allow integrations to add routes to projects, that ship have prebuild, plug-n-play functionality, pages, or route handlers. For consumers of such integrations, this would allow highly plug-and-play like experiences for adding functionality.

![image](https://user-images.githubusercontent.com/17054057/170668303-d9eb0197-d672-4fbe-b1c7-c496a1dd405f.png)

## Proposed API

```ts
export interface InjectedRoute {
  pattern: string,
  entryPoint: string
}

function injectRoute(injectRoute: InjectedRoute): void {};
```

### Usage

In your integration, you could then use the API like so:

```js
export default defineConfig({
  integrations: [
    {
      name: 'my-netlify-integration',
      hooks: {
        'astro:config:setup': ({injectRoute}) => {
          injectRoute({
            /** The route on which to output the entryPoint */
            pattern: '/admin',
            /** Bare module specifier pointing to a pre-made admin page */
            entryPoint: 'my-netlify-integration/admin.astro'
          })
        }
      }
    }
  ]
});
```

## Usecases

Some usecases for this could be:
- Adding an `/admin` page for headless CMSes
- Implementation of [tailwind-config-viewer](https://github.com/rogden/tailwind-config-viewer)
- Authentication providers could very easily ship the required redirectCallback routes etc, e.g. `googleProvider()`, `facebookProvider()`
- Plug and play payment integrations, endpoints, webhooks, all setup via a single integration, e.g. `paypal()`, or `stripe()`

## Open questions/discussion

Over the weekend I worked on an implementation for this proposal for which I created a draft PR [here](https://github.com/withastro/astro/pull/3457).

Some discussion:
In Fred original message he proposed a `{ pattern: '/tailwind/*' }`. In my implementation, I decided against using `/*` as dynamic routing patterns, because it differs from Astro's built-in routing conventions, e.g.: `[dynamic]`, or `[...spread]`, and I didnt want to introduce a new way of handling routes for consistency.

Additionally, file extensions in the `pattern` are not required, because the route pattern is only used for the resulting route, or where the `entryPoint` will get output to. This means there's some discrepancy between the provided `pattern` and the `entryPoint`, e.g. an injected route's `pattern` may be dynamic, but the actual file will always 'just' be either an `.astro` file or a `.js` file;

```js
injectRoute({
  /** Note how the file extension is not present in the pattern */
  pattern: '/foo/dynamic/[dynamic]',
  /** Note how the dynamic brackets are not present in the entryPoint */
  entryPoint: 'foo/dynamic.astro'
})
```

This is somewhat of a discrepancy between Astro's regular authoring experience, where for dynamic routes, you would write `/foo/[dynamic].astro`. 

This is only an 'issue' however for integration-authors, but may require some good documentation. Consumers of integrations that use `injectRoute` should not run into this.

For completeness sake, here are some examples of route configurations I've tested in my PR:
```js
injectRoute({
  pattern: '/foo',
  entryPoint: 'foo/index.js'
})

injectRoute({
  pattern: '/foo/bar',
  entryPoint: 'foo/bar.js'
})

injectRoute({
  pattern: '/foo/baz',
  entryPoint: 'foo/baz.astro'
})

injectRoute({
  pattern: '/foo/dynamic/[dynamic]',
  entryPoint: 'foo/dynamic.astro'
})

injectRoute({
  pattern: '/foo/spread/[...spread]',
  entryPoint: 'foo/spread.astro'
})
```

And this is what these routes will get turned into during the build:

```js
{
  type: 'endpoint',
  pattern: /^\/foo$/,
  params: [],
  component: 'node_modules/foo/index.js',
  pathname: '/foo'
},
{
  type: 'endpoint',
  pattern: /^\/foo\/bar$/,
  params: [],
  component: 'node_modules/foo/bar.js',
  pathname: '/foo/bar'
},
{
  type: 'page',
  pattern: /^\/foo\/baz\/?$/,
  params: [],
  component: 'node_modules/foo/baz.astro',
  pathname: '/foo/baz'
},
{
  type: 'page',
  pattern: /^\/foo\/dynamic\/([^/]+?)\/?$/,
  params: [ 'dynamic' ],
  component: 'node_modules/foo/dynamic.astro',
  pathname: undefined
},
{
  type: 'page',
  pattern: /^\/foo\/spread(?:\/(.*?))?\/?$/,
  params: [ '...spread' ],
  component: 'node_modules/foo/spread.astro',
  pathname: undefined
}
```

**Comment by stevenle** [comment]
Thanks for this RFC. This perfectly captures a project i'm working on that injects a `/cms` route, design token values (e.g. tailwind in your example), and authentication.

One thing I might suggest is: in addition to defining an `entryPoint` that points to a single astro file, maybe allow for a way to have it point to a folder that has a `pages/` routing structure. This might mean an entire path would be taken over by the `injectRoute` but it allows developers of an integration to be able to structure routes the same way they would develop any other astro project.

**Comment by aFuzzyBear** [comment]
The implications of this particular feature is huge, creating routes is a bit hacky at the present moment. 

Having seen some varied implementations of doing this, I would welcome the use of a standardised way of injecting premade routes into the manifest some way.

I don't think this is tangential to the conversation but discussing unlocking private routes `_route` should be also included within the scope. 

This way it would allow for `_admin` routes and projects such as `astro-gui` to create phenomenal private developer experiences.

**Comment by natemoo-re** [comment]
Notes from call:

- Conflict handling is an important consideration. Integrations should provide configuration options for the routes that are mounted, Astro should throw an Error if a conflict is detected.
- We should instruct authors of integrations that _only_ run in SSR to warn or error.
- How are dev-only routes handled? Up to the integration author.

Next steps:
- Turn this discussion into a PR! We'll resolve any discussion points there and move forward.

**Reply by natemoo-re** [reply]
- @FredKSchott mentioned Partytown integration which uses `sirv`. Possibly dogfood this integration with `@astrojs/partytown`?

**Reply by natemoo-re** [reply]
- Underscores in routes should be supported in `pattern`? Good to check whether this works or not here.

**Comment by thepassle** [comment]
Notes from Astro Community Call:

Q: What about route collisions?
A: add some error handling (throw err when duplicate routes)

Q: Pointing to entire folders, e.g.: `pages/`
A: preferably not, currently using built-in `require.resolve` functionality, 'fishing' on the FS is less reliable (thinking of package export maps etc)

TODO:
- turn into proper RFC
- docs for conventions around adding route patterns (e.g.: using `_`)
- document dev-only routes `if(command === 'dev') injectRoute(routeConfig);`

**Comment by tanepiper** [comment]
This would solve a lot of issues around dynamic page generation on change detection, allowing for an incremental rebuild.

I wonder if something like the below example would be possible (also as a JSON or YAML) where you can pass a set of routes to regenerate only a set of pages.  I suppose it would be up to Astro to figure out if it touches anything else, and most likely a sitemap rebuild too, etc

*routes.js*
```js
export default [
 { route: '/page/page1', target: '/content/pages/page1.astro' },
 { route: '/page/page2', target: '/content/pages/page2.md' },
 { route: '/docs/[dynamic]', target: '/content/docs/[dynamic].astro', props: [{dynamic: 'api1'}, {dynamic: 'api2' }] }, // could be async source 
 { route: /old/path', target: '/new/path', redirect: 301, reason: 'page path has updated' }
]
```

Pass it as an optional property in the CLI:

```bash
astro build --routes=routes.js
astro build --routes=routes.js --onlyRoutes // Only build the routes, don't update any pages not mentioned
```

I don't really want Astro to care too much about file storage and cache invalidation :D At least not yet
